lib/snippets/kubernetes-deploy

```diff
@@ -0,0 +1,172 @@
+
+require 'open3'
+require 'securerandom'
+require 'erb'
+require 'json'
+require 'yaml'
+require 'shellwords'
+require 'tempfile'
+require 'logger'
+
+class KubernetesDeploy
+  class FatalDeploymentError < StandardError; end
+
+  def initialize(namespace:, environment:, current_sha:, template_folder: nil, context:)
+    @namespace = namespace
+    @context = context
+    @current_sha = current_sha
+    @template_path = File.expand_path('./' + (template_folder || "config/deploy/#{environment}"))
+    # Max length of podname is only 63chars so try to save some room by truncating sha to 8 chars
+    @id = current_sha[0...8] + "-#{SecureRandom.hex(4)}" if current_sha
+  end
+
+  def run
+    validate_configuration
+    set_kubectl_context
+    validate_namespace
+    apply_all_templates
+  rescue FatalDeploymentError => error
+    logger.fatal(error.message)
+    exit 1
+  end
+
+  def template_variables
+    {
+      'current_sha' => @current_sha,
+      'deployment_id' => @id,
+    }
+  end
+
+  private
+
+  def validate_configuration
+    errors = []
+    if ENV["KUBECONFIG"].nil? || !File.file?(ENV["KUBECONFIG"])
+      errors << "Kube config not found at #{ENV["KUBECONFIG"]}"
+    end
+
+    if @current_sha.nil?  || @current_sha.empty?
+      errors << "Current SHA must be specified"
+    end
+
+    if !File.directory?(@template_path)
+      errors << "Template path #{@template_path} doesn't exist"
+    elsif Dir.entries(@template_path).select {|file| file =~ /\.yml(.erb)?$/}.size.zero?
+      errors << "#{@template_path} doesn't contain valid templates (postfix .yml or .yml.erb)"
+    end
+
+    if @namespace.nil? || @namespace.empty?
+      errors << "Namespace must be specified"
+    end
+
+    if @context.nil? || @context.empty?
+      errors << "Context must be specified"
+    end
+
+    raise FatalDeploymentError, "Configuration invalid: #{errors.join(", ")}" if errors.length.nonzero?
+    logger.info("All required parameters and files are present")
+  end
+
+  def apply_all_templates
+    found = 0
+    Dir.foreach(@template_path) do |file|
+      file_path = "#{@template_path}/#{file}"
+      if File.extname(file) == '.yml'
+        found += 1
+        apply_template(file_path)
+      elsif File.extname(file) == '.erb'
+        found += 1
+        render_and_apply_template(file_path)
+      end
+    end
+    raise FatalDeploymentError, "No templates found in #{@template_path}" if found.zero?
+  end
+
+  def render_and_apply_template(file_path)
+    erb_template = ERB.new(File.read(file_path))
+    erb_binding = binding
+    template_variables.each do |var_name, value|
+      erb_binding.local_variable_set(var_name, value)
+    end
+    content = erb_template.result(erb_binding)
+
+    f = Tempfile.new(['kube_template', '.yml'])
+    f.write(content)
+    f.close
+    apply_template(f.path, original_path: file_path)
+  ensure
+    f.unlink if f
```

Comment:
I have been in the habit of explicitly unlinking anyway because of this recommendation in the Tempfile docs:

> When a Tempfile object is garbage collected, or when the Ruby interpreter exits, its associated temporary file is automatically deleted. This means that's it's unnecessary to explicitly delete a Tempfile after use, though it's good practice to do so: not explicitly deleting unused Tempfiles can potentially leave behind large amounts of tempfiles on the filesystem until they're garbage collected. The existence of these temp files can make it harder to determine a new Tempfile filename. Therefore, one should always call unlink or close in an ensure block

I'm guessing this script's ruby process is really short-lived though, making this extra-unnecessary here?